addons/account_edi_ubl/models/account_edi_format_en_16931.py

```diff
@@ -0,0 +1,243 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import models, fields, _
+from odoo.exceptions import UserError
+from odoo.tests.common import Form
+
+
+COUNTRY_EAS = {
+    'HU': 9910,
+
+    'AD': 9922,
+    'AL': 9923,
+    'BA': 9924,
+    'BE': 9925,
+    'BG': 9926,
+    'CH': 9927,
+    'CY': 9928,
+    'CZ': 9929,
+    'DE': 9930,
+    'EE': 9931,
+    'UK': 9932,
+    'GR': 9933,
+    'HR': 9934,
+    'IE': 9935,
+    'LI': 9936,
+    'LT': 9937,
+    'LU': 9938,
+    'LV': 9939,
+    'MC': 9940,
+    'ME': 9941,
+    'MK': 9942,
+    'MT': 9943,
+    'NL': 9944,
+    'PL': 9945,
+    'PT': 9946,
+    'RO': 9947,
+    'RS': 9948,
+    'SI': 9949,
+    'SK': 9950,
+    'SM': 9951,
+    'TR': 9952,
+    'VA': 9953,
+
+    'SE': 9955,
+
+    'FR': 9957
+}
+
+
+# This edi_format is "abstract" meaning that it provides an additionnal layer for similar edi_format (formats deriving from
+# EN16931) that share some functionnalities, but needs to be extended to be used.
+class AccountEdiFormat(models.Model):
+    _inherit = 'account.edi.format'
+
+    ####################################################
+    # Helpers
+    ####################################################
+
+    def _check_en_16931_invoice_configuration(self, invoice):
+        errors = []
+
+        # partners
+        customer = invoice.commercial_partner_id
+        if not self._get_ubl_partner_values(customer).get('en_16931_endpoint', False):
+            errors.append(_('Customer need to have an endpoint id, typically this is the vat number. Please note that countries outside of EU are not supported.'))
+
+        supplier = invoice.company_id.partner_id.commercial_partner_id
+        if not self._get_ubl_partner_values(supplier).get('en_16931_endpoint', False):
+            errors.append(_('Supplier need to have an endpoint id, typically this is the vat number. Please note that countries outside of EU are not supported.'))
+
+        # invoice line
+        if invoice.invoice_line_ids.filtered(lambda l: not (l.product_id.name or l.name)):
+            errors.append(_('Each invoice line must have a product or a label.'))
+
+        if invoice.invoice_line_ids.tax_ids.invoice_repartition_line_ids.filtered(lambda r: r.use_in_tax_closing) and \
+           not supplier.vat:
+            errors.append(_("When vat is present, supplier must have a vat number."))
+
+        return errors
+
+    def _get_ubl_partner_values(self, partner):
+        # OVERRIDE
+        values = super()._get_ubl_partner_values(partner)
+        if partner.country_id.code in COUNTRY_EAS:
```

Comment:
I would prefer to initialize both values (en_16931_endpoint and en_16931_endpoint_scheme) only if the scheme exists, its safer if one those values is checked to check if the partner has an EN16931 endpoint. If this point is valid, I would need to do a `if` anyway, so I think this can stay this way.